### ISSUE TITLE
links.csv and sentence logs have inconsistent view of which links exist

### ISSUE THREAD

**Issue by Yorwba** [Tatoeba/tatoeba2#1724]
Grepping for [sentence 929679](https://tatoeba.org/sentences/show/929679) in `links.csv` yields three links (and their mirror images):
```
464908	929679
507792	929679
673271	929679
```
But [sentence 67321](https://tatoeba.org/sentences/show/673271) no longer exists (which is how I noticed that something was off) and in the logs on the sentence page you can see Horus unlinking it and adding [sentence 543362](https://tatoeba.org/sentences/show/543362). But that link is not in `links.csv`.

Assuming `links.csv` is really just a database dump, it seems that at some point Horus added log messages about changing links without actually changing them (possible cause: the sentences were linked to each other before one of them was edited and turned into a duplicate of the other).

One way to fix this would be the script in docs/database/scripts/clean_ghost_links.sql, but that'd simply delete the invalid links without restoring the missing ones. Is there some record of the replacement for a deleted sentence in the database or would one have to parse Horus' comments to fix up the links?

Also, I don't know whether the Horus bug that caused the inconsistency has been fixed or not. #304 seems to have been closed without having discovered the root cause of the ghost sentences.

**Comment by trang** [comment]
Thanks for the analysis @Yorwba!

You're diving into an issue that is very likely caused by a [major incident we had back in June 2017](https://blog.tatoeba.org/2017/06/tatoeba-is-back.html).

Taking your example with  sentence `#673271` that has been deleted but still has a link in `links.csv`, what I assume happened is the folowing:

- Horus deleted `#673271` in on April 24, 2017.
- Our SSD died two months later, on June 11, and we lost all our data.
- We restored data from our backup which seemingly was only a snapshot from February 19, 2017. On this backup, Horus didn't delete `#673271` yet.
- We then used the CSV files generated on June 10 and June 3 to restore whatever we could that was missing.
- The easiest thing to do was to restore any addition of data. We did that and brought Tatoeba back online. At this point, sentence `#673271` would still be in the database, but the logs would be restored up to June 3, so they would indicate that `#673271` has been deleted.
- Then we worked on restoring modifications and deletion of sentences (#1484). At this point, sentence `#673271` was effectively removed from the database. We can find it in the list of deleted sentences that I posted in [one of my last comments](https://github.com/Tatoeba/tatoeba2/issues/1484#issuecomment-354617885).
- The links were not restored yet (#1561), which is why the pair 673271-929679 is still hanging around in the database.


Now to answer your questions.

> One way to fix this would be the script in docs/database/scripts/clean_ghost_links.sql, but that'd simply delete the invalid links without restoring the missing ones. Is there some record of the replacement for a deleted sentence in the database or would one have to parse Horus' comments to fix up the links?

You would have to parse the `contributions.csv` (it's not a file listed on the Downloads page but you can see it from the [export folder](https://downloads.tatoeba.org/exports/)). Although we know there are some inconsistencies in our logs (#1653), it's still reliable enough.

The `contributions.csv` file is generated with [the following SQL query](https://github.com/Tatoeba/tatoeba2/blob/dev/docs/database/scripts/weekly_exports.sql#L61):
```
SELECT u.username, c.datetime, c.action, c.type, c.sentence_id,
       c.sentence_lang, c.translation_id, c.text       
FROM contributions c LEFT JOIN users u ON c.user_id = u.id
ORDER BY c.datetime ASC
```

With the following values for `action` and `type`:
- action = `insert`, `update` or `delete`
- type = `link`, `sentence` or `license`


> I could write a script to restore all the missing links using my own user account, but I'd like official permission before doing something like that.

I think would be best to generate files with the list of links to re-delete and with the list of links to re-add. Then we can run a script directly on the database to fix all that.


> Also, is there an easy way to check whether a sentence is missing from sentences.csv because it was marked as unreliable and not because it was deleted?

I could generate for you a CSV file with all the sentences. But for a more generic solution, you would have anyway to check against the `contributions.csv`.

- If a link is present in `links.csv` but the last log entry for this link has `action=delete`, then it should be deleted.
- And vice versa, if the last log for a link has `action=insert`, but this link is not present in `links.csv`, then it should be re-added.

**Comment by Yorwba** [comment]
Thank you for the explanation. I had already seen the incident referenced somewhere, but it wasn't clear to me that some data hasn't been recovered yet despite not being totally lost. I guess that makes this a duplicate of #1561.

Your hint regarding the `contributions.csv` file was also very helpful. Unfortunately, the logs therein do not reliably indicate which sentences really exist. There are 14825 sentences with a deletion record that are not actually deleted, like [#5893](https://tatoeba.org/sentences/show/5893). There are also 34394 sentences which were deleted without any log of that happening, like [#333](https://tatoeba.org/sentences/show/333).

Even weirder, there are five sentences with an unlink log that has `\N` in place of the id of the translation being unlinked: [#380193](https://tatoeba.org/sentences/show/380193), [#258203](https://tatoeba.org/sentences/show/258203), [#560921](https://tatoeba.org/sentences/show/560921), [#343669](https://tatoeba.org/sentences/show/343669) and [#320383](https://tatoeba.org/sentences/show/320383)

The uncertainty regarding which sentences actually exist also makes it hard to tell whether a link is legitimate or not. For example, [#166073](https://tatoeba.org/sentences/show/166073) is linked to [#248432](https://tatoeba.org/sentences/show/248432) according to logs, but the latter has been deleted in favor of [#22781](https://tatoeba.org/sentences/show/22781) without that being recorded in the logs, while still updating the links.

Generally, the problem is that sometimes the logs are right and the database would need to be updated to fit the logs, and sometimes the database is right and the logs are simply missing.

In cases where both logs and database agree on the existence of a sentence, it should however be possible to change links without messing up. If a sentence definitely doesn't exist, links to it can be deleted safely. If both sentences definitely exist and the logs say that they're linked, they should probably be linked, unless the unlink log was lost.

**Comment by trang** [comment]
> I guess that makes this a duplicate of #1561.

Sort of, but #1561 is a bit more narrow, since it's only about restoring what has been lost from the incident. There are likely inconsistencies outside of that time frame. I will close #1561 and keep your issue instead.

> Your hint regarding the contributions.csv file was also very helpful. Unfortunately, the logs therein do not reliably indicate which sentences really exist.
> ...
> Generally, the problem is that sometimes the logs are right and the database would need to be updated to fit the logs, and sometimes the database is right and the logs are simply missing.

I see your point. 

So I generated for you a CSV file with [all the sentences](https://downloads.tatoeba.org/exports/all_sentences_20181127.tar.bz2) along with the [contributions.csv](https://downloads.tatoeba.org/exports/contributions_20181127.tar.bz2) so you can have the latest logs.

The sentences were exported with the following query:
```
SELECT id, lang, text, correctness, created, modified FROM `sentences`
```
We only use `0` and `-1` for the correctness. Sentences with `correctness=-1` are not exported in the official files.

**Comment by Yorwba** [comment]
I've thought a bit more about this and I think there are essentially four ways that the logs can diverge from the database content.

1. A user wants to change something, but only the logs are written, the actually intended change does not happen.

2. A user wants to change something and it works, but the logs are missing.

3. Dirty merge: It seems like in the early days of the deduplication script, logs were simply merged together, leading to sentences with multiple creation times and outdated link numbers.

4. The crash: changes after the backup date are recorded in the logs, but missing from the database if they weren't restored.

In case 1, the user is likely going to notice that nothing changed and will simply try again until it works. That should also apply to Horus, I think. I expect there to be only few instances of case 1.

In case 2, nobody is going to notice anything, so there will be plenty of those around.

Case 3 is actually similar, in that the naive merging of logs differs from the current implementation by not having a bunch of unlinks followed by relinks. So that is a case of missing logs just like 2.

Case 4 is probably the largest single cause, and where the logs are more accurate than the database.

Now here's a bit of a gamble: logs have been getting more accurate over time, especially regarding case 3, which doesn't happen anymore. (Right? Most instances I've seen were from 2011 or so.) So we can try to trust the logs for everything that happened after the backup date and assume the database is correct for everything before that.

Based on that assumption, I've gone through `contributions_20181127.csv`, ignored all lines before 2017-02-18 (using a padding of one day to the backup date because I don't know the exact time) and then recorded additions and deletions of links afterwards. Then I diffed them against the `links.csv` file and put them into two files, one containing [2949 links that should be deleted, but that are still in `links.csv`](https://raw.githubusercontent.com/Yorwba/tatoeba-ghost-links/master/data/output/replayed_links_to_delete.csv) and one with [131153 links that aren't in `links.csv`, but should be](https://raw.githubusercontent.com/Yorwba/tatoeba-ghost-links/master/data/output/replayed_links_to_restore.csv). So, essentially what you said about using the action of the last log entry to decide, but only if that entry happened after the backup date.

I manually verified a random sample (drawn using `sort -R ... | head` so I wasn't just looking at events close in time) of ten links from each file. The list of links to be deleted seems correct, they all were either to sentences that had been deleted, or there were comments on the sentences requesting them to be unlinked. The list of links to restore contains some that point to sentences created after Saturday's `links.csv`, some that were affected by the crash, but also some that point to sentences that were deleted without correctly removing the links. I thought about filtering those out with the list of all sentences you provided (thanks!), but #1631 means that even that list is incomplete. When those missing sentences are restored, we'd have to do the link restoration procedure again, unless we add them now and accept that that means keeping links to correctly deleted sentences around for a bit longer.

The next step would be to address #1631 by reconstructing the missing sentences from `contributions.csv` the same way as the links. Once that is done, any links to deleted sentences can be deleted once and for all.

At that point, the database would be caught up on the events recorded in the logs, and it would be nice to update the logs to reflect whatever the current state is. I'd hesitate to remove any of the logged information (what if it turns out that restoration _still_ wasn't done completely?), but appending the missing contributions should be harmless.

**Comment by trang** [comment]
> Now here's a bit of a gamble: logs have been getting more accurate over time, especially regarding case 3, which doesn't happen anymore. (Right? Most instances I've seen were from 2011 or so.)

Sounds about right. We introduced Horus some time in 2014. Before that, deduplication was done with [this procedure](https://github.com/Tatoeba/tatoeba2/blob/dev/docs/database/procedures/cleaning_up_duplicates.sql), which didn't log any change into the `contributions` table.

> Then I diffed them against the links.csv file and put them into two files, one containing 2949 links that should be deleted, but that are still in links.csv and one with 131153 links that aren't in links.csv, but should be. 

Interesting, there's a much larger amount of missing links than I expected. Thanks a lot for doing this work. I'll ask other members of the community to have a look at these lists and if no one finds any issue, I can apply the deletion and restoration of links.

> The list of links to restore contains some that point to sentences created after Saturday's links.csv, some that were affected by the crash, but also some that point to sentences that were deleted without correctly removing the links. I thought about filtering those out with the list of all sentences you provided (thanks!), but #1631 means that even that list is incomplete. 

I didn't really look into #1631 until now, but there's a chance that when I tried to restore sentences from the CSV files, not all the sentences created after Feb 19 were restored. 

Here's the ZIP containing all the CSV files we had from 2017-06-10: https://drive.google.com/open?id=0B3fWO5pH-xU1VDRva2RLZE8wODQ
The sentence `#6103407` exists in the `sentences_detailed.csv` so it should have been restored, but apparently wasn't. Or was somehow deleted without log.

Here are the scripts I used to restore data from the CSV:
https://gist.github.com/trang/f7e4b8c8860d32d58c1e51c253c89c18
https://gist.github.com/trang/9365403a6da60e8cd150dec55ae300f0

In any case we can use the `contributions` to restore the missing sentences. But you can use the data up there for better analysis of the problem if you wish. Also I'm thinking if some sentences are missing, it could be that some logs are missing as well. I didn't check that.

> At that point, the database would be caught up on the events recorded in the logs, and it would be nice to update the logs to reflect whatever the current state is. I'd hesitate to remove any of the logged information (what if it turns out that restoration still wasn't done completely?), but appending the missing contributions should be harmless.

If we restore the data based on the logs, the logs will be reflecting the current state. What would there be to update?

**Comment by trang** [comment]
I went ahead and restored the links listed in [replayed_links_to_restore.csv](https://raw.githubusercontent.com/Yorwba/tatoeba-ghost-links/master/data/output/replayed_links_to_delete.csv) and removed the links in [replayed_links_to_delete.csv](https://raw.githubusercontent.com/Yorwba/tatoeba-ghost-links/master/data/output/replayed_links_to_restore.csv).

Then I removed ~16,000 of the links that were restored because they actually became ghost links due to deduplication.
- https://gist.github.com/trang/91fb205d93de6e8fb501a99c76464409
- https://gist.github.com/trang/51a904b8ad0455b9c453bd08a3aed962

I will close this issue now. 

We certainly still have some inconsistency between what's in the links.csv and what's in the logs but at least a large amount of missing links have been restored.

We also still have a bunch of ghost links hanging around. It seems a lot of them are due to Horus not deleting the links when deleting a duplicate sentence. Will not remove them for now as they're not causing much trouble.
